File: miyata/miyata-reading-circle-tdd/src/main/java/money/Money.java
package money;

public abstract class Money {

    protected int amount;
    protected String currency;

    Money(int amount, String currency) {
        this.amount = amount;
        this.currency = currency;
    }

    abstract Money times(int multiplier);
    public String currency() {
        return currency;
    };

    public boolean equals(Object object) {
        if (object instanceof Money money) {
            return amount == money.amount
                    && getClass().equals(money.getClass());
        }
        return false;
    }
    static Money dollar(int amount) {
        return new Dollar(amount, "USD");
    }
    static Money franc(int amount) {
        return new Franc(amount, "CHF");
    }
}

File: murnana/Unity-TestDrivenDevelopmentByExample/Assets/TestDrivenDevelopmentByExample/Scripts/Runtime/Money.cs
using System;

namespace TDD
{
    public class Money : IEquatable<Money>,
                         IExpression
    {
        /// <summary>
        /// 通貨
        /// </summary>
        private readonly string m_Currency;

        /// <summary>
        /// 通貨の数
        /// </summary>
        public int Amount { get; }

        public Money(int amount, string currency)
        {
            Amount     = amount;
            m_Currency = currency;
        }

        public static Money Dollar(int amount)
        {
            return new Money (amount: amount, currency: "USD");
        }

        public static Money Franc(int amount)
        {
            return new Money (amount, "CHF");
        }

        public IExpression Times(int multiplier)
        {
            return new Money (amount: Amount * multiplier, currency: m_Currency);
        }

        public string Currency()
        {
            return m_Currency;
        }

        #region Implementation of IExpression

        /// <inheritdoc />
        public Money Reduced(Bank bank, string to)
        {
            var rate = bank.GetRate(from: m_Currency, to: to);
            return new Money(amount: Amount / rate, currency: to);
        }

        /// <inheritdoc />
        public IExpression Plus(IExpression added)
        {
            return new Sum(augend: this, addend: added);
        }

        #endregion

        #region Equality members

        /// <inheritdoc />
        public bool Equals(Money other)
        {
            if (other == null)
            {
                return false;
            }

            return (Currency() == other.Currency())
                && (Amount == other.Amount);
        }

        #endregion

        #region Overrides of Object

        /// <inheritdoc />
        public override string ToString()
        {
            return $"{Amount} {m_Currency}";
        }

        #endregion
    }
}
